value.go

```diff
@@ -67,45 +68,213 @@ func getValue(info TypeInfo, v mapping.Value) (any, error) {
+		if vo.Kind() == reflect.Ptr {
+			if vo.IsNil() {
+				return nil
+			}
+			v = vo.Elem().Interface()
+			continue
+		}
+		return v
+	}
+}
+
+func isNil(i any) bool {
+	if i == nil {
+		return true
+	}
+
+	value := reflect.ValueOf(i)
+	kind := value.Kind()
+
+	switch kind {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Slice:
+		return value.IsNil()
+	}
+	return false
+}
+
+func createValueByReflection(v any) (Type, *mapping.Value, error) {
+	t := TYPE_INVALID
+	switch vv := v.(type) {
+	case nil:
+		t = TYPE_SQLNULL
+	case bool:
+		t = TYPE_BOOLEAN
+	case int8:
+		t = TYPE_TINYINT
+	case int16:
+		t = TYPE_SMALLINT
+	case int32:
+		t = TYPE_INTEGER
+	case int64:
+		t = TYPE_BIGINT
+	case int:
+		t = TYPE_BIGINT
+		v = int64(vv)
+	case uint8:
+		t = TYPE_UTINYINT
+	case uint16:
+		t = TYPE_USMALLINT
+	case uint32:
+		t = TYPE_UINTEGER
+	case uint64:
+		t = TYPE_UBIGINT
+	case uint:
+		t = TYPE_UBIGINT
+		v = uint64(vv)
+	case float32:
+		t = TYPE_FLOAT
+	case float64:
+		t = TYPE_DOUBLE
+	case string:
+		t = TYPE_VARCHAR
+	case []byte:
+		t = TYPE_VARCHAR
+		v = string(vv)
+	case time.Time:
+		t = TYPE_TIMESTAMP
+	}
+	if t != TYPE_INVALID {
+		return t, createValueByTypeId(t, v), nil
+	}
+	if ss, ok := v.(fmt.Stringer); ok {
+		return TYPE_VARCHAR, createValueByTypeId(TYPE_VARCHAR, ss.String()), nil
+	}
+	if isNil(v) {
+		t = TYPE_SQLNULL
+		return t, createValueByTypeId(t, v), nil
+	}
+	r := reflect.ValueOf(v)
+	switch r.Kind() {
+	case reflect.Ptr:
+		return createValueByReflection(getPointerValue(v))
+	case reflect.Slice:
+		vv, err := getMappedSliceValueByReflection(r.Interface(), false)
+		return TYPE_LIST, vv, err
+	case reflect.Array:
+		vv, err := getMappedSliceValueByReflection(r.Interface(), true)
+		return TYPE_ARRAY, vv, err
 	default:
-		return nil, unsupportedTypeError(reflect.TypeOf(v).Name())
+		return t, nil, unsupportedTypeError(reflect.TypeOf(v).Name())
 	}
+}
 
-	return &vv, err
+func getMappedSliceValueByReflection[T any](val T, isArray bool) (*mapping.Value, error) {
+	createFunc := mapping.CreateListValue
+	if isArray {
+		createFunc = mapping.CreateArrayValue
+	}
+	vSlice, err := extractSlice(val)
+	if err != nil {
+		return nil, fmt.Errorf("could not cast %T to []any: %s", val, err)
+	}
+	var childValues []mapping.Value
+	defer destroyValueSlice(childValues)
+	if len(vSlice) == 0 {
+		lt := mapping.CreateLogicalType(TYPE_SQLNULL)
+		defer destroyLogicalTypes(toPtr([]mapping.LogicalType{lt}))
+		return toPtr(createFunc(lt, childValues)), nil
+	}
+	elementType := TYPE_INVALID
+	for _, v := range vSlice {
+		et, vv, err := createValueByReflection(v)
+		if err != nil {
+			return nil, err
+		}
+		if elementType == TYPE_INVALID || elementType == TYPE_SQLNULL {
+			elementType = et
+		}
+		childValues = append(childValues, *vv)
+	}
+	if elementType == TYPE_INVALID {
+		return nil, unsupportedTypeError(reflect.TypeOf(val).Name())
+	}
+	lt := mapping.CreateLogicalType(elementType)
```

Comment:
It is not sufficient to only return the child `Type` here - `mapping.CreateLogicalType` only works for primitive types. For a `LIST` containing another `LIST`, we need to use `CreateListType`, for example. `typeInfo` does that, if set correctly: `func (info *typeInfo) logicalType() mapping.LogicalType`.